js/ui/workspaceMonitor.js

```diff
@@ -10,276 +10,142 @@ const WorkspaceMonitor = new Lang.Class({
-    },
+    _minimizeWindowCompleted: function(shellwm, actor) {
+        let app = this._windowTracker.get_window_app(actor.meta_window);
 
-    // The sequence of events is
-    // Window is mapped but not yet in system: _windowAdded
-    // Window is newly created: _windowAdded -> _mapWindow
-    // Window is minimized: _minimizeWindow
-    // Window is unminimized: _mapWindow
-    // Window is closed: _destroyWindow -> _windowRemoved
-    // Minimized window is closed: _windowRemoved
-    //
-    // This allows us to separate the work
-    // _windowAdded: Handles tracking of known windows
-    //               checks if a window is interesting and adds it to our list
-    //               of known windows. Also checks if it is minimized and if so
-    //               adds it to the list of minimized windows
-    // _windowRemoved: Is the opposite of _windowAdded
-    //
-    // _mapWindow: Handles tracking of visible windows
-    // _minimizeWindow:
-    // _destroyWindow: These are the opposite of _mapWindow for the 2 different
-    //                 ways a window can be taken off the screen: minimizing and
-    //                 closing
-    _realWindowAdded: function(metaWindow) {
-        // We special-case eos-speedwagon here, as we don't want to go
-        // back to the overview if the only window that was dismissed was the
-        // splash screen.
-        // See also shell_app_is_interesting_window() in src/shell-app.c, which
-        // was added for the same purpose.
-        // This code eventually should be changed to track applications instead
-        // of windows.
-        if (!Shell.WindowTracker.is_window_interesting(metaWindow) ||
-            metaWindow.get_role() == 'eos-speedwagon') {
-            return false;
+        if (app in this._trackedApps) {
+            this._setAppVisible(app, this._appHasVisibleWindows(app));
         }
 
-        this._addKnownWindow(metaWindow);
-        return true;
+        this._updateOverview();
     },
 
-    _windowAdded: function(metaWorkspace, metaWindow) {
-        this._realWindowAdded(metaWindow);
-    },
+    _unminimizeWindow: function(shellwm, actor) {
+        let app = this._windowTracker.get_window_app(actor.meta_window);
 
-    _windowRemoved: function(metaWorkspace, metaWindow) {
-        // Remove the window from our known windows
-        // Window will not be in knownWindows if it wasn't
-        // interesting to us.
-        if (!this._windowIsKnown(metaWindow)) {
-            return;
+        if (app in this._trackedApps ) {
+            this._setAppVisible(app, this._appHasVisibleWindows(app));
         }
-
-        this._removeKnownWindow(metaWindow);
     },
 
-    _windowGoingInvisible: function() {
-        // Check if we're minimizing the very last window
-        if (this.visibleWindows == 1) {
-            Main.layoutManager.prepareForOverview();
+    _updateOverview: function() {
+        if (this._visibleApps.size == 0) {
```

Comment:
(optional suggestion): I wonder if instead of always keeping track of the visible apps it wouldn't be simpler to compute it on demand from the tracked applications at the time when `_updateOverview()` is called.